### PR TITLE
Convert type_id from String to a node

### PR DESCRIPTION
**Pull request by jansul**
This PR modifies the `type_id`/`type_id!` methods of the parser to return a new node (`Ast::TypeId`) rather than return a `String`. Any other node affected by this has also been updated.

This should make it easier to implement language server features where it can be useful to exact positions of some nodes. e.g. the `name` of an `Ast::Component`

I've tested this change against the `mint-realworld` repository, running `mint docs` etc and everything seemed OK, but could possible use more testing given how much is changed.

### CODE REVIEW

File: src/ls/hover.cr [hunk 1]
```diff
@@ -58,10 +58,11 @@ module Mint
             parent = stack[1]?
 
             case node
-            when Ast::Variable
-              # If the first node under the cursor is a variable then
-              # get the associated nodes information and hover that
-              # otherwise get the hover information of the parent.
+            when Ast::Variable, Ast::TypeId
+              # If the first node under the cursor is a `Ast::Variable`
+              # or `LSP::TypeId`, then get the associated nodes
```
Comment (gdotdesign) [inline src/ls/hover.cr:63]: It's not `Ast::TypeId`?

**Review by gdotdesign — APPROVED** [review]
Looks good :+1: I guess if it compiles and works with the real word repo it should be good! Once merged I'll test `mint-ui` with it.

File: src/parsers/html_body.cr [hunk 2]
```diff
@@ -40,6 +40,8 @@ module Mint
           case tag
           when Ast::Variable
             tag.value
+          when Ast::TypeId
+            tag.value
```
Comment (Sija) [inline src/parsers/html_body.cr:44]:
```suggestion
          when Ast::Variable, Ast::TypeId
            tag.value
```

File: src/type_checkers/use.cr [hunk 3]
```diff
@@ -22,7 +22,7 @@ module Mint
 
       # This is checked by the provider so we assume it's there
       subscription =
-        records.find!(&.name.==(provider.subscription))
+        records.find!(&.name.==(provider.subscription.value))
```
Comment (Sija) [inline src/type_checkers/use.cr:25]: `#name` is not a `TypeId`?

```suggestion
        records.find!(&.name.value.==(provider.subscription.value))
```

**Review by Sija — COMMENTED** [review]
I'm wondering, wouldn't be cleaner to implement `TypeId#to_s` and `Comparable(String)` to avoid explicit usage of `#value` property?

**Comment by gdotdesign** [comment]
> I'm wondering, wouldn't be cleaner to implement `TypeId#to_s` and `Comparable(String)` to avoid explicit usage of `#value` property?

I guess we could do that, however we currently don't, so I don't know if it's in the scope of this PR or not :thinking:

**Comment by Sija** [comment]
@gdotdesign I'm pretty sure we could at least implement `TypeId#==(other: TypeId)` to avoid doin' in manually everywhere, like:

```cr
ast.stores.find(&.name.value.==(connect.store.value))
```

File: src/type_checkers/use.cr (hunk 3 above)
Comment (jansul) [reply src/type_checkers/use.cr:25]: Actually not in this case :smile:!  `records` here is a list of `TypeChecker::Record`'s, where `name` is a [`String`](https://github.com/mint-lang/mint/blob/master/src/type_checker/record.cr#L8)

**Comment by jansul** [comment]
I was very tempted to implement those methods, however as it's not something we are doing on any other node (especially with `Ast::Variable` where we also use `.value`) I thought best to do it like this for consistency.

Happy to make any changes :smile:  `TypeId#==(other: TypeId)` would probably be the safest and least confusing. 

Just noticed a bug where I'm not using `.value` when passing to errors so I will fix that too ...!
image

**Comment by jansul** [comment]
Fixed all of the above :smile:  

As an aside it may be worth hardening the [`bold` method of `BlockBuilder`](https://github.com/mint-lang/mint/blob/master/src/message.cr#L19) to only accept strings the same as `text` as that would have caught that issue!

**Review by Sija — APPROVED** [review]

**Pull request merged**